### ISSUE TITLE
How to produce a boxplot with groups, where the groups are side-by-side

### ISSUE THREAD

**Issue by bisraelsen** [JuliaPlots/StatPlots.jl#56]
I am making a box plot grouped by a certain variable. By default the boxes appear overlapping on each other. Is there a way to have them side-by-side to avoid the overlap?

**Comment by mkborregaard** [comment]
This is much easier to adress if you post an MWE and a plot. 
The short answer, though, is that this cannot be done automatically at the moment. But that it'd be nice to have. If you do post an MWE I might suggest a workaround.

**Comment by kescobo** [comment]
@mkborregaard is problem related to #44? I'm suspecting I will need this functionality more in the nearish future. I managed to figure out a work-around, and I'd be willing to make an attempt to formalize it into a PR if that would be helpful (or to take over #45 if that's a thing that makes sense...)

**Comment by mkborregaard** [comment]
It is, yes. See also https://github.com/JuliaPlots/StatPlots.jl/issues/72 where we're currently discussing this. It'd be great to resolve this now :-)

**Comment by kescobo** [comment]
Ahh... I read that one, but didn't realize it was related but it seems clear now. Looks like ~ 4/7 currently open issues are related to this... good that the discussion is being had.

**Comment by piever** [comment]
I'm also trying to come up with a solution for #44  (and this sort of group related issues). As noted already, the key of the problem is [here](https://github.com/JuliaPlots/Plots.jl/blob/master/src/series.jl#L513), where `group` is making the series. What it should do, instead, is to rearrange the `y` argument into a Matrix or Array of Arrays.

When the `groupedbar` recipe is being processed, the only available information at the time of grouping is the type of `args` (in this case it's a `GroupedBar`, with one field, itself called `args` containing the tuple of the arguments of the plot call). If we want to keep it a type recipe, the only solution that came to my mind is the following:

Add in `RecipesBase` a function:

`group_as_matrix(t) = false`

Then [here](https://github.com/JuliaPlots/Plots.jl/blob/master/src/series.jl#L513) it should check `group_as_matrix(args)` and if it's false, it groups normally with the series, otherwise it does a  Matrix. After that, to specify that we want the `groupedbar` to be grouped as a Matrix, we add a dispatch in StatPlots:

`RecipesBase.group_as_matrix(t::GroupedBar) = true`

**Comment by mkborregaard** [comment]
Why would that function need to be in RecipesBase?

**Comment by mkborregaard** [comment]
With regards to "groupedbar" I am currently leaning towards not using the "group" system in Plots, but to leave it as a user recipe as it is, and then add two keywords as earlier suggested by @kescobo , e.g. "bargroup" and "colorgroup", that controlled these plots. It's not a problem with extra keywords for a well-documented user recipe.

**Comment by mkborregaard** [comment]
With regards to boxplots side-by-side (the original question) I'd once again need to see an MWE to completely understand what the issue is (I asked @bisraelsen for that in my first post) - but if it's just to group boxplots by a variable you can pass that as x:
```
boxplot(rand(["a", "b", "c"], 100), randn(100))
```
<img width="579" alt="skaermbillede 2017-08-23 kl 22 42 22" src="https://user-images.githubusercontent.com/8429802/29637417-5a34817a-8854-11e7-9bb9-8c8f48384482.png">

**Comment by kescobo** [comment]
@mkborregaard To be honest, I'm a bit intimidated by wading into the guts of this package. I think I more or less understand what's happening in the `bar.jl` file - would I just add some logic to re-order the array if the keywords are present?

Also, how do I check if the keywords are present? :-/

**Comment by mkborregaard** [comment]
Try to play around with it and see if you can get it to behave :-) Inside the recipe, you have a Dict called `d` in the namespace, all keyword attributes to Plots are keys in that Dict. So things like `get(d, :colorgroup, nothing)` work.

**Comment by ValdarT** [comment]
Might not be the best place to raise this issue but currently the default behavior of boxplot when given only one array is weird. If `bs = randn(100)` then `boxplot(bs)` treats every point as a separate category. I think it should behave like `boxplot([""], bs)` instead.

Just mentioning here in case this this turns into a PR. Could change that as well while at it. : )

**Comment by mkborregaard** [comment]
👍

**Comment by ValdarT** [comment]
Another minor issue. The boxplots are too wide, which is especially bad if there is only one.
Interestingly enough, they are leaner when using the group parameter. 

Compare these two
```
using StatPlots, RDatasets; iris = dataset("datasets", "iris")

boxplot(iris, :Species, :PetalLength)
boxplot(iris, :Species, :PetalLength, group=:Species)
```

Mentioning here for the same reason as before.

**Comment by mkborregaard** [comment]
You can adjust the width of boxplots with the `bar_width` attribute.

**Comment by ValdarT** [comment]
Thanks. I know that I can control the box width (although it doesn't actually have an effect in case of only one box), I was just pointing out that the default behavior is arguably not the best and is different (and better) when `group` is specified. 
I hope to find time over the weekend to see if I can improve on the 2 things I've mentioned here.

**Comment by mkborregaard** [comment]
👍 to changing the behaviour with only one box. WRT the width, 80% of the diff is a Plots-wide standard. You can try argueing your case with a demonstration of box plots most commonly used in the literature, but there's a likelihood we won't agree with you :-)

**Comment by piever** [comment]
Closed by #192
